### PR TITLE
Fix end date of all-day events in iCal imports

### PR DESCRIPTION
**Pull request by charludo**
### Proposed changes

- simply subtract a day from the imported date if the event is all-day

### Side effects

Even though it is probably not likely to occur "in the wild", we previously had the possibility of *only* either the start or end date being a `datetime.date` instead of a `datetime.datetime`.

This could lead to either all-day events with and end time, or non-all-day events with no end time.
I opted to change the logic slightly, and now events are all-day as soon as either the start or end has no time.

### Resolved issues

Fixes: #3109

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (david-venhoff) [inline integreat_cms/cms/utils/external_calendar_utils.py:38]: Apparently we have been doing this wrong the entire time :see_no_evil: 
More context: https://github.com/digitalfabrik/integreat-cms/pull/3075#discussion_r1801247908

Github won't let me make a suggestion, but the fix would be to move this line to the first line of the function

**Review by david-venhoff — APPROVED** [review]
Thanks!

Looks good to me

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (JoeyStk) [inline integreat_cms/cms/utils/external_calendar_utils.py:85]: Do you think we can put this into it's own method to make the method a bit smaller? :)

**Review by JoeyStk — APPROVED** [review]
Nice, thank you very much. That fixes the issue! :)
I only have one small nit :)

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (charludo) [reply integreat_cms/cms/utils/external_calendar_utils.py:85]: I originally wanted to say "yes, sure!", but - that function would either have to take `start` and `end` as input and return `start_date`, `start_time`, `end_date`, `end_time`, `is_all_day` as a giant 5-tuple, or be at least three different functions.

Not sure if that is wirth it? 😅  If you insist I'll implement it, I'm just a bit ambivalent here. Small functions are nice, but I am not sure it would help with readability.

Comment (JoeyStk) [reply integreat_cms/cms/utils/external_calendar_utils.py:85]: Oops, I didn't realize I was blocking this PR :/ To be honest, at least for now, I don't have a clear opinion or idea. To me this is another function that keeps growing and eventually will be hard to maintain. However I also see your point and you're absolutely right. So I don't know :shrug:

Comment (charludo) [reply integreat_cms/cms/utils/external_calendar_utils.py:85]: > To me this is another function that keeps growing and eventually will be hard to maintain.

That's absolutely valid. The only thing I can really give as a counterargument is that it's not a short function, but it's also not a complex function - no loops, and this is the only if/else branch. *So far* it's OK to maintain, I think :D

But I'll leave it up to you.  If the above is not convincing, LMK, and I'll use this PR to try and refactor the function.

Comment (JoeyStk) [reply integreat_cms/cms/utils/external_calendar_utils.py:85]: > So far it's OK to maintain, I think :D
Yeah, on second thought I agree. I think we can merge this PR :)

**Pull request merged**
